### PR TITLE
Java Fluent SDK

### PR DESCRIPTION
**Pull request by andrewmchen**
## What changes are proposed in this pull request?
 
In this PR, we propose adding a "fluent" API for Java. This API allows users to manage MLflow runs similar to how they would in the Python SDK. For example, 

```java
    RunInfo runCreated = client.createRun(expId);
    System.out.println("CreateRun: " + runCreated);
    String runId = runCreated.getRunUuid();

    // Log parameters
    client.logParam(runId, "min_samples_leaf", "2");
    client.logParam(runId, "max_depth", "3");

    // Log metrics
    client.logMetric(runId, "auc", 2.12F);
    client.logMetric(runId, "accuracy_score", 3.12F);
    client.logMetric(runId, "zero_one_loss", 4.12F);

    // Update finished run
    client.setTerminated(runId, RunStatus.FINISHED);
```

changes to

```java
    MlflowContext mlflow = new MlflowContext();
    ActiveRun run = mlflow.startRun("run");
    run.logParam("alpha", "0.0");
    run.logMetric("MSE", 0.0);
    run.setTags(ImmutableMap.of(
      "company", "databricks",
      "org", "engineering"
    ));
    run.endRun();
```

It will also support adding `mlflow.runName`, `mlflow.user`, `mlflow.source.type`, and `mlflow.parentRunId` tags on `startRun` similarly to how we do in the Python SDK.
## How is this patch tested?
 
(Details)
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [x] API 
- [ ] REST-API 
- [x] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [x] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by andremesarovic** [comment]
Using AutoCloseable idiom would provide a better experience similar to our current Python context manager idiom.
Example:
```
new RunContext(mlflowClient, expId, sourceName) {
  println(s"Run ID: ${getRunId()}")
}
```

**Comment by andremesarovic** [comment]
Using the AutoCloseable idiom would provide a better experience similar to our current Python context manager idiom.
Example:
```
new RunContext(mlflowClient, expId) {
  println(s"Run ID: ${getRunId()}")

  logParam("alpha","0.5")
  logMetric("rmse",0.786)
  setTag("fluent","true")
  setTag("mlflow.note.content","my **bold** note")

  val now = new java.util.Date()
  new PrintWriter("/tmp/info.txt") { write(s"Info: $now") ; close }
  logArtifact(new File("/tmp/info.txt"),"info")

  val dir = new File("/tmp/run_artifacts")
  dir.mkdirs()
  new PrintWriter(s"$dir/info1.txt") { write(s"Info1 at: $now") ; close }
  logArtifacts(dir)
  logArtifacts(dir,"dir")
}
```

**Comment by andrewmchen** [comment]
Is the code you provided in Scala? What would it look like in Java? Also are there docs for that sort of usage in scala?

One thing we didn't like about the Autocloseable interface was that AFAIK there isn't a way to see in the implementation of `close()` whether or not an exception has been thrown.

**Comment by andremesarovic** [comment]
Here are Scala and Java examples: 

https://github.com/amesar/mlflow-fun/blob/master/mlflow-java/README.md#runcontext-fluent-proposal

**Comment by andrewmchen** [comment]
Thanks for taking the time to prototype those. Super helpful. 

I couldn't actually find docs on how to use the try-with-resources that way in Scala. What keywords do you need to google to find docs on that type of syntax?

Also, it looks like your implementation of `RunContext.close` will always set the RunStatus to FINISHED. That's inconsistent with python which will set it to FAILED if it throws an exception. Is there a way to get that behavior with the run context syntax?

**Comment by andremesarovic** [comment]
> your implementation of RunContext.close will always set the RunStatus to FINISHED

It was just a quick prototype. Have added setRunStatus() - https://github.com/amesar/mlflow-fun/blob/master/mlflow-java/src/main/java/org/mlflow/tracking/RunContext.java#L60

> Python run status.  Is there a way to get that behavior with the run context syntax?

Good point, Python traps an exception. Not sure there is an easy way to do the same in Java - needs research. You could always put a try/catch and then call setRunStatus (ugly).

> I couldn't actually find docs on how to use the try-with-resources that way in Scala. What keywords do you need to google to find docs on that type of syntax?

Scala surprisingly doesn’t have a “with-resources” idiom like Java. I googled all over in vain. I found some ad-hoc implementations out there - none jumped out as super intuitive. In fact, using the Java version in Scala seemed simpler than the Scala ones! BTW, did the research 9 months ago.

One of them:  https://www.phdata.io/try-with-resources-in-scala

TryWith.run(new RunContext2(mlflowClient, expId))(_.close) { run =>
  run.logParam("alpha","val1")
  run.logMetric("m1",0.123)
}

**Comment by andremesarovic** [comment]
Further research: apparently when the block in Scala is executed, the close() method is not called upon block exit. In Java it is called correctly.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/ActiveRun.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  public void logParam(String key, String value) {
+    client.logParam(getId(), key, value);
+  }
+
+  public void setTag(String key, String value) {
+    client.setTag(getId(), key, value);
+  }
+
+  public void logMetric(String key, double value) {
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/ActiveRun.java]: We should document the implied step and timestamp.

Comment (aarondav) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/ActiveRun.java]: Same with other overloaded methods, where we pick some values by default

File: mlflow/java/client/src/main/java/org/mlflow/tracking/ActiveRun.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  public void setTags(Map<String, String> tags) {
+    List<RunTag> protoTags = tags.entrySet().stream().map((tag) ->
+      RunTag.newBuilder().setKey(tag.getKey()).setValue(tag.getValue()).build()
+    ).collect(Collectors.toList());
+    client.logBatch(getId(), Collections.emptyList(), Collections.emptyList(), protoTags);
+  }
+
+  public void logArtifact(Path localPath) {
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/ActiveRun.java]: We should probably generally document all these methods to a similar extent as `fluent.py`. We should expect that Java users may only read the documentation for this file (i.e., that they wouldn't go to `MlflowClient`'s docs.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+   */
+  public String getExperimentId() {
+    return this.experimentId;
+  }
+
+  /**
+   * Starts a MLflow run. To log data to newly created MLflow run see the methods on
+   * {@link ActiveRun}. MLflow runs should be ended using {@link ActiveRun#endRun()}
+   *
+   * @param runName The name of this run.
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Maybe mention that this is used for display purposes only, and is stored in the `mlflow.runName` tag.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+   * Like {@link #startRun(String)} but sets the {@code mlflow.parentRunId} tag in order to create
+   * nested runs.
+   *
+   * @param runName The name of this run.
+   * @param parentRunId The ID of this run's parent
+   * @return An {@code ActiveRun} object to log data to.
+   */
+  public ActiveRun startRun(String runName, String parentRunId) {
+    Map<String, String> tags = new HashMap<>();
+    tags.put(MlflowTagConstants.RUN_NAME, runName);
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Should we allow a null runName too?

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    }
+    RunInfo runInfo = client.createRun(createRunBuilder.build());
+
+    ActiveRun newRun = new ActiveRun(runInfo, client);
+    return newRun;
+  }
+
+  /**
+   * Like {@link #startRun(String)} but will terminate the run after the activeRunFunction is
+   * executed.
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Can you write example code with this? I assume there's a fancy lambda-y thing you can do in Java.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/samples/FluentExample.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package org.mlflow.tracking.samples;
+
+import com.google.common.collect.ImmutableMap;
+import org.mlflow.tracking.ActiveRun;
+import org.mlflow.tracking.MlflowContext;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public class FluentExample {
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/samples/FluentExample.java]: Like the examples!

File: mlflow/java/client/src/main/java/org/mlflow/tracking/utils/DatabricksContext.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      return null;
+    }
+    return new DatabricksContext(configProvider);
+  }
+
+  /**
+   * May return null if replId was not set on the configProvider. This will happen with DBR
+   * @return
+   */
+  public String getReplId() {
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/utils/DatabricksContext.java]: Remove?

File: mlflow/java/client/src/test/java/org/mlflow/tracking/MlflowContextTest.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  }
+
+  @Test
+  public void testStartRun() {
+    // Sets the appropriate tags
+    ArgumentCaptor<CreateRun> createRunArgument = ArgumentCaptor.forClass(CreateRun.class);
+    MlflowContext mlflow = setupMlflowContext();
+    mlflow.setExperimentId("123");
+    mlflow.startRun("apple", "parent-run-id");
```
Comment (aarondav) [inline mlflow/java/client/src/test/java/org/mlflow/tracking/MlflowContextTest.java]: Could we also test `startRun` with the Consumer, as well as testing every public method?

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java (hunk 4 above)
Comment (andrewmchen) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Sure

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java (hunk 5 above)
Comment (andrewmchen) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Good idea

File: mlflow/java/client/src/main/java/org/mlflow/tracking/ActiveRun.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    client.logArtifact(getId(), localPath.toFile());
+  }
+
+  /**
+   * Uploads the given local file to the run's root artifact directory. For example,
+   *
+   *   <pre>
+   *   activeRun.logArtifact("/my/localModel", "model")
+   *   mlflowClient.listArtifacts(activeRun.getId(), "model") // returns "model/localModel"
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/ActiveRun.java]: Nice example!

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+  public MlflowClient getClient() {
+    return this.client;
+  }
+
+  /**
+   * Sets the experiment to log runs to by name.
+   * @param experimentName the name of the experiment to log runs to.
+   * @throws IllegalArgumentException if the experiment name does not match an existing experiment
+   */
+  public void setExperimentName(String experimentName) throws IllegalArgumentException {
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Let's have these return `MLflowContext` to allow call-chaining.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.function.Consumer;
+
+/**
+ * Main entrypoint used to start MLflow runs to log to. This is a higher level interface than
+ * {@code MlflowClient} and provides convenience methods to keep track of active runs and to set
+ * default tags on runs which are created through {@code MlflowContext}
+ *
+ * On construction, MlflowContext will choose a default experiment ID to log to depending on your
+ * environment. To log to a different experiment, use {@link #setExperimentId(String)} or
+ * {@link #setExperimentName(String)}
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Maybe we can include a usage example here as well.

**Review by aarondav — APPROVED** [review]

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+import java.util.*;
+import java.util.function.Consumer;
+
+/**
+ * Main entrypoint used to start MLflow runs to log to. This is a higher level interface than
+ * {@code MlflowClient} and provides convenience methods to keep track of active runs and to set
+ * default tags on runs which are created through {@code MlflowContext}
+ *
+ * On construction, MlflowContext will choose a default experiment ID to log to depending on your
```
Comment (aarondav) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowContext.java]: Add `<p>` and period to this paragraph and previous

**Pull request merged**